### PR TITLE
ci: treewide: deny used_underscore_binding

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

Using `_`-prefixed variables are odd: https://github.com/solana-labs/solana/pull/31239#discussion_r1169860871

#### Summary of Changes

Remove and deny them in the future.

Fixes #

### CODE REVIEW

File: ledger/src/blockstore_processor.rs
Comment (ryoqun) [inline ledger/src/blockstore_processor.rs:446]: cc: @apfitzge @taozhu-chicago as per https://github.com/solana-labs/solana/pull/31239#discussion_r1169860871

File: core/src/window_service.rs
Comment (ryoqun) [inline core/src/window_service.rs:295]: @carllin is this needed still? seems this is added at https://github.com/solana-labs/solana/pull/10109#discussion_r1174807289

File: core/src/ancestor_hashes_service.rs
Comment (ryoqun) [inline core/src/ancestor_hashes_service.rs:1042]: it seems that it's fine to remove this cfg(test) code added here: https://github.com/solana-labs/solana/pull/18912#discussion_r1174814547

File: core/tests/snapshots.rs
Comment (ryoqun) [inline core/tests/snapshots.rs:76]: i think this comment properly communicates the implicit intention explicitly, which is added https://github.com/solana-labs/solana/pull/29942#discussion_r1174834165

File: ci/test-checks.sh
Comment (apfitzge) [inline ci/test-checks.sh:81]: Just adding a comment here with ref to clippy-lint description: https://rust-lang.github.io/rust-clippy/master/#used_underscore_binding

File: tokens/src/main.rs
Comment (apfitzge) [inline tokens/src/main.rs:38]: nit: For me personally, I think the naming `exit_for_ctrlc_handler` could imply that it is separate from the other exit signal?

I think the following is essentially what the original code was intending to do?
```suggestion
    // Initialize CTRL-C handler to ensure db changes are written before exit.
    ctrlc::set_handler({
        let exit = exit.clone();
        move || {
            exit.store(true, Ordering::SeqCst);
        }
    })
```

**Review by apfitzge — DISMISSED** [review]
lgtm - just get approval from the others for the actual removed vars/fields before merging

File: tokens/src/main.rs
Comment (ryoqun) [reply tokens/src/main.rs]: yeah, that confusion is valid concern: https://github.com/solana-labs/solana/pull/31319/commits/9de98b256734642c9e5c2866e28816363b791c48

(btw, seems I can't automatically apply your suggestion on deleted lines...: 

>  image
)

File: core/tests/snapshots.rs
Comment (ilya-bobyr) [reply core/tests/snapshots.rs:76]: **minor**

Using underscored names to keep references to objects that need to live for some time is one of the useful use cases for underscored names.
So, if this object is needed to keep the `TempDir` around, using an underscore name seems exactly right.
A comment is also helpful, of course.
I think `#[allow(dead_code)]` is a less common choice for this.
So I suggest reverting to the underscored name, but keeping the comment.
```suggestion
    // Keep TempDir::drop from running to retain that dir for the duration of test.
    _accounts_tmp_dir: TempDir,
```

A linter should then catch if someone will actually start using this field.
Which would not happen with `#[allow(dead_code)]`.

File: core/tests/snapshots.rs
Comment (ryoqun) [inline core/tests/snapshots.rs:133]: here

File: core/tests/snapshots.rs
Comment (ryoqun) [inline core/tests/snapshots.rs:131]: heer

File: core/tests/snapshots.rs
Comment (ilya-bobyr) [reply core/tests/snapshots.rs:76]: I should have checked that move is considered "use".
Thank you for the explanation.
I think what you originally did is then perfectly reasonable.

Thinking about it for a few minutes, this seems reasonable as well:
```rust
    _accounts_tmp_dir: accounts_tmp_dir,
```
If it keeps clippy happy.  It can be read as that `accounts_tmp_dir` is used in this function, but it is then not accessed at all, except for the Drop call.

File: tokens/src/main.rs
Comment (apfitzge) [reply tokens/src/main.rs]: > (btw, seems I can't automatically apply your suggestion on deleted lines...:

you just don't want to give me credit on the commit! 😿

File: core/tests/snapshots.rs
Comment (ryoqun) [reply core/tests/snapshots.rs]: > Thinking about it for a few minutes, this seems reasonable as well:
> 
> ```rust
>     _accounts_tmp_dir: accounts_tmp_dir,
> ```

oh, that's good to hear from you. accordingly, I've changed my mind: 6347179921d513aee2664762f494496a1d7b0a5f

I really dislikes `allow(...)`. :)

File: core/src/ancestor_hashes_service.rs
Comment (carllin) [reply core/src/ancestor_hashes_service.rs:1042]: Yeah this seems fine to remove given that the `AncestorHashesService` does not shut down if `duplicate_slots_reset_sender` send fails

File: core/src/window_service.rs
Comment (carllin) [reply core/src/window_service.rs:295]: yeah i think we can remove this

**Comment by ryoqun** [comment]
@yihau could you approve this quickly? I've finished all the stamps for this tree-wide small changes.

**Review by ilya-bobyr — APPROVED** [review]

**Pull request merged**
